### PR TITLE
MattT/CASEFLOW-4869 Create EMO 'Return to BVA Intake' Task

### PR DESCRIPTION
**Pull request by ThorntonMatthew**
Resolves [CASEFLOW-4869](https://vajira.max.gov/browse/CASEFLOW-4869)

### Description
Create a task action and an accompanying modal to allow a member of the EMO organization to return an appeal to BVA Intake if it was sent to them in error.

### Acceptance Criteria
- [x] Code compiles correctly
- [x] "Return to Board Intake" task action is created
- [x] "Return to Board Intake" modal is created
- [x] Modal allows for the user to enter a description on why the appeal is been sent back to BVA Intake
- [x] Instructions are not optional per designs
- [x] EMO can successfully send a predocket appeal to BVA Intake
- [x] Upon success, show the appropriate success banner
- [x] Upon fail, show a failure (most likely a Slim Error Alert in the modal)
- [x] After assignment the EMO task has a status of "completed"
- [x] The BVA Intake PreDocketTask should be updated from "on_hold" to "assigned"and the appeal shows in the BVA Intake "Ready for Review" tab
- [x] Whenever BVA Intake sends an appeal back to the EMO, the previous task should not appear in the EMO's queue any longer. Only the most recent `EducationDocumentSearchTask` should be visible.

#### Plan For 4869-Specific Changes

0. You will want to run the frontend and backend of Caseflow separately for some of this testing. `make run-frontend` and `make run-backend` to run each, respectively.

1. Send an appeal with an education-related issue to pre-docket by running the following command in the Rails console (`make c`):

```
# Whenever prompted for a CSS_ID, feel free to use BVAISHAW, or any other Intake user's CSS_ID.
appeal = FactoryBot.create(:education_document_search_task, :assigned).appeal
```
Running `appeal.treee` should yield the following task tree:

Task tree before returning the appeal

2. Switch to an EMO user. Go to to the EMO organizations queue page and be sure you are on the unassigned tab. [Click here](http://localhost:3000/organizations/edu-emo?tab=education_unassigned) to go there (if you are running Caseflow locally on port 3000).

3. Click on the 'Veteran Name (Veteran File Number)' link in the 'Case Details' column for the appeal that was created in step one.

4. On the appeal's Queue page, click on the Actions dropdown and select 'Return to Board Intake'. A modal should appear on the screen.

5. Click on the 'Return' button without placing any text in the text field in the modal. Ensure that the "Text field cannot be blank." error appears, and that the form cannot be submitted.

6. Stop your backend. Enter some text into the text field in the modal and attempt click the 'Return' button. The following error should appear at the top of the modal: "There was an error processing your request. Please retry your action and contact support if errors persist."

7. Start back up Caseflow's backend, then click the 'Return' button in the modal.

8. You should be redirected to the EMO's unassigned tab (instead of the current user's assigned tab, which was the previous behavior). Ensure that the appeal that was returned to intake is no longer present in this tab.

9. Select the 'Assigned' tab on the EMO organization's Queue page. The appeal should be listed here. It will continue to be shown on this tab until BVA Intake completes (dockets) or cancels the PreDocketTask within this appeal's task tree.

10. `appeal.treee` should yield the following task tree now:

Task tree after successfully returning an appeal to Intake

11. Switch to a BVA Intake user.

12. Navigate to the BVA Intake organization's Queue page. Go to the 'Ready for Review' tab. [Click here](http://localhost:3000/organizations/bva-intake?tab=bvaReadyForReview) to navigate there. The appeal should appear here.

13. Feel free to docket the appeal. You can then confirm that it goes into Bva Intake's completed tab, and that it no longer appears in either of the EMO's unassigned or assigned tabs. Once the EMO's completed tab is available in the feature branch, that is where it will appear.

The output of `appeal.treee` after docketing the appeal:

Task tree of an appeal that has been docketed

14. Create another appeal in such a way that mimics whenever the EMO has sent an appeal to BVA Intake.

`appeal2 = FactoryBot.create(:education_document_search_task, :completed).appeal`

15. Switch to an Intake user and navigate to appeal2's Queue page.

16. This time, select "Return appeal to Education Service" from the actions dropdown.

17. Switch one last time to an EMO user. Go to the [EMO org's queue page](http://localhost:3000/organizations/edu-emo?tab=education_unassigned), and ensure that appeal2's task appears in the unassigned tab, and not the assigned one.

`appeal2.treee`'s output:

Task tree of an appeal sent back to EMO

#### Plan for CodeClimate Fixes

1. You will want to ensure that the EMO's unassigned tab contains the correct columns-- especially if the Veteran Documents column is needing to be shown.

2. Make sure that VeteranRecordRequestTasks are created if an education appeal, or any appeal, isn't sent to pre-docket:

veteran-record

3. Any appeal being sent to pre-docket should not have a VeteranRecordRequestTask in its task tree.

- [ ] For higher-risk changes: [Deploy the custom branch to UAT to test](https://github.com/department-of-veterans-affairs/appeals-deployment/wiki/Applications---Deploy-Custom-Branch-to-UAT)

### User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

#### Happy Path

happy-path

The appeal's task tree after completion of the Happy Path:

appeal_successfully_sent_to_intake

#### No Context Provided

no-input-error

#### Server Failure

server-failure

### Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### CODE REVIEW

File: client/constants/TASK_ACTIONS.json [hunk 1]
```diff
@@ -1,417 +1,417 @@
...
+    "label": "Send to Board Intake",
+    "value": "modal/vha_send_to_board_intake",
+    "func": "vha_send_to_board_intake"
+  },
+  "VHA_MARK_TASK_IN_PROGRESS": {
+    "label": "Mark task in progress",
+    "value": "modal/mark_task_in_progress",
+    "func": "vha_mark_task_in_progress"
+  },
+  "EMO_RETURN_TO_BOARD_INTAKE": {
```
Comment (ThorntonMatthew) [inline client/constants/TASK_ACTIONS.json:412]: "EMO_RETURN_TO_BOARD_INTAKE" is the portion that has been added in this PR. For some reason the prettier lint check in the CircleCI workflow kept rejecting this file, which is why I "prettified" it.

File: app/models/queue_tabs/education_emo_assigned_tasks_tab.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  def description
+    format(COPY::EDUCATION_EMO_QUEUE_PAGE_ASSIGNED_TASKS_DESCRIPTION, assignee.name)
+  end
+
+  def active_parents_of_emo_tasks_assigned_to_rpos_or_bva
+    (on_hold_tasks + closed_tasks).map(&:parent).reject(&:closed?)
+  end
+
+  def task_ids_without_newer_siblings
```
Comment (ThorntonMatthew) [inline app/models/queue_tabs/education_emo_assigned_tasks_tab.rb:22]: These changes were necessitated due to an oversight of mine in [4600](https://github.com/department-of-veterans-affairs/caseflow/pull/17186). I failed to account for the new `EducationDocumentSearchTask` that is created upon BVA Intake sending an appeal back to the EMO, and that any previous EMO tasks for the same appeal need to not be shown.

File: spec/factories/task.rb [hunk 3]
```diff
@@ -505,7 +505,7 @@ def self.find_first_task_or_create(appeal, task_type, **kwargs)
       end
 
       factory :education_document_search_task, class: EducationDocumentSearchTask do
-        parent { create(:pre_docket_task, appeal: appeal) }
+        parent { create(:pre_docket_task, appeal: appeal, assigned_to: BvaIntake.singleton) }
```
Comment (ThorntonMatthew) [inline spec/factories/task.rb:508]: This was a fix to allow navigation to `/queue/appeal/#{appeal.uuid}`. Without proper organization assignments the task tree isn't able render properly, and an error is thrown.

File: spec/feature/queue/pre_docket_spec.rb [hunk 4]
```diff
@@ -379,7 +375,11 @@ def bva_intake_dockets_appeal
 
         expect(page).to have_content("Review Documentation")
         find(".cf-select__control", text: COPY::TASK_ACTION_DROPDOWN_BOX_LABEL).click
-        find("div", class: "cf-select__option", text: Constants.TASK_ACTIONS.EMO_SEND_TO_BOARD_INTAKE_FOR_REVIEW.label).click
+        find(
```
Comment (ThorntonMatthew) [inline spec/feature/queue/pre_docket_spec.rb:378]: This line was split up to satisfy the CodeClimate/rubocop line length limit check.

File: app/models/appeal.rb [hunk 5]
```diff
@@ -339,7 +339,7 @@ def vha_has_issues?
   end
 
   def edu_predocket_needed?
-    request_issues.active.any? { |ri| ri.benefit_type == "education" && ri.is_predocket_needed }
+    request_issues.active.any? { |ri| ri.benefit_type == "education" && ri.predocket_needed? }
```
Comment (ThorntonMatthew) [inline app/models/appeal.rb:342]: This change was made to resolve a CodeClimate issue where it wanted `is_predocket_needed` to be `predocket_needed?` instead. There is still a CC issue here due to `ri` being referenced twice, however I am still thinking of how best to fix this.

File: app/models/queue_tabs/education_emo_unassigned_tasks_tab.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  # In each case, if the EMO task is assigned, then it should meet each use case for the unassigned tab
+  def tasks
+    Task.includes(*task_includes).visible_in_queue_table_view.where(assigned_to: assignee).active
+  end
+
+  # Column names defined in each tab in education predocket due to different columns needed
+  # Actual names found in QUEUE_CONFIG.json file
+  def column_names
+    if show_reader_link_column
```
Comment (ThorntonMatthew) [inline app/models/queue_tabs/education_emo_unassigned_tasks_tab.rb:28]: COLUMN_NAMES was split out to resolve a CC issue.

**Review by jtsangVA — APPROVED** [review]
lgtm!

**Pull request merged**